internal/services/notification/runeventssender.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2023 Sorint.lab
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package notification
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"sort"
+	"strconv"
+	"sync"
+	"testing"
+
+	"github.com/gorilla/mux"
+	"github.com/rs/zerolog"
+	"github.com/sorintlab/errors"
+
+	"agola.io/agola/internal/testutil"
+	"agola.io/agola/internal/util"
+	"agola.io/agola/services/runservice/types"
+)
+
+type runEventsSender struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	started bool
+
+	log zerolog.Logger
+
+	t *testing.T
+
+	mu sync.Mutex
+
+	listenAddress string
+	exposedURL    string
+
+	runEvents *runEvents
+
+	config *runEventsSenderConfig
+}
+
+type runEventsSenderConfig struct {
+	skipAfterRunEventSequence bool
```

Comment:
Is this really needed. It should behave like the real runservice and it doesn't have such option.

Reply:
@sgotti I added this behavior to test the expected error `run event %d already processed`

Reply:
1. It's the notification service that could ask for an event that was concurrently already processed. Not the runservice returning events not requested.
2. An already processed event should not return an error but just ignore it.